tests/hash_table_test.c

```diff
@@ -60,7 +60,132 @@ static int test_hash_table_put_get_fn(struct aws_allocator *alloc, void *ctx) {
+                             "Returned value for %s, should have been %s", "tweedle dee", "tweedle dum");
+
+    ret = aws_hash_table_find(&hash_table, (void *)key_2, &pElem);
+    ASSERT_SUCCESS(ret, "Hash Map get should have succeeded.");
+    ASSERT_BIN_ARRAYS_EQUALS("what's for dinner?", strlen("what's for dinner?"), aws_string_bytes(pElem->key),
+                             ((struct aws_string *)pElem->key)->len,
+                             "Returned key for %s, should have been %s", "what's for dinner?", "what's for dinner?");
+    ASSERT_BIN_ARRAYS_EQUALS("deadbeef", strlen("deadbeef"), aws_string_bytes(pElem->value),
+                             ((struct aws_string *)pElem->value)->len,
+                             "Returned value for %s, should have been %s", "what's for dinner?", "deadbeef");
+
+    ret = aws_hash_table_find(&hash_table, (void *)key_3, &pElem);
+    ASSERT_SUCCESS(ret, "Hash Map get should have succeeded.");
+    ASSERT_BIN_ARRAYS_EQUALS(bytes, 10, aws_string_bytes(pElem->key),
+                             ((struct aws_string *)pElem->key)->len,
+                             "Returned key for %02hhx%02hhx%02hhx%02hhx%02hhx%02hhx%02hhx%02hhx%02hhx%02hhx should have been same",
+                             bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7], bytes[8], bytes[9]);
+    ASSERT_BIN_ARRAYS_EQUALS("hunter2", strlen("hunter2"), aws_string_bytes(pElem->value),
+                             ((struct aws_string *)pElem->value)->len,
+                             "Returned value for binary bytes should have been %s", "hunter2");
+    aws_hash_table_clean_up(&hash_table);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_hash_table_string_clean_up, test_hash_table_string_clean_up_fn)
+static int test_hash_table_string_clean_up_fn(struct aws_allocator *alloc, void *ctx) {
+    /* Verify that clean up happens properly when a destructor function is used only on keys or only on values. */
+    struct aws_hash_table hash_table;
+    struct aws_hash_element *pElem;
+    int was_created;
+
+    const struct aws_string * key_1 = aws_string_from_literal_new(alloc, "Once upon a midnight dreary,");
+    AWS_STATIC_STRING_FROM_LITERAL(val_1, "while I pondered, weak and weary,");
+    const struct aws_string * key_2 = aws_string_from_literal_new(alloc, "Over many a quaint and curious");
+    AWS_STATIC_STRING_FROM_LITERAL(val_2, "volume of forgotten lore--");
+    const struct aws_string * key_3 = aws_string_from_literal_new(alloc, "While I nodded, nearly napping,");
+    AWS_STATIC_STRING_FROM_LITERAL(val_3, "suddenly there came a tapping,");
+
+    const struct aws_string * dyn_keys[] = {key_1, key_2, key_3};
+    const struct aws_string * static_vals[] = {val_1, val_2, val_3};
+
+    int ret = aws_hash_table_init(&hash_table, alloc, 10, aws_hash_string, aws_string_eq,
+                                  aws_string_destroy, NULL); /* destroy keys not values */
```

Comment:
I assume this is implicitly tested via the debug allocator

Reply:
Not quite sure what you mean here. I haven't used the debug allocator, but I could look into adding some more tests with it. I did run all of these tests through valgrind to make sure that all allocated memory was being freed.

Reply:
I understand now that the memory tracking allocator is running by default on the tests. I verified that if the destructor function was not set properly on the init, this test will not pass because of memory not being freed. So yes, the test framework allocator is working properly to help us on this test.